cache/ringbuffer.go

```diff
@@ -14,181 +14,240 @@
 	head, tail, size int
+	revisionOf       RevisionOf[T]
 }
 
-// batch groups all events that share one ModRevision.
-type batch struct {
-	rev    int64
-	events []*clientv3.Event
+type entry[T any] struct {
+	revision int64
+	items    []T
 }
 
-type KeyPredicate = func([]byte) bool
-
-type IterFunc func(rev int64, events []*clientv3.Event) bool
+type (
+	KeyPredicate      = func([]byte) bool
+	RevisionOf[T any] func(T) int64
+	IterFunc[T any]   func(rev int64, items []T) bool
+)
 
-func newRingBuffer(capacity int) *ringBuffer {
+func newRingBuffer[T any](capacity int, revisionOf RevisionOf[T]) *ringBuffer[T] {
 	// assume capacity > 0 – validated by Cache
-	return &ringBuffer{
-		buffer: make([]batch, capacity),
+	return &ringBuffer[T]{
+		buffer:     make([]entry[T], capacity),
+		revisionOf: revisionOf,
 	}
 }
 
-func (r *ringBuffer) Append(events []*clientv3.Event) {
+func (r *ringBuffer[T]) Append(items []T) {
 	start := 0
-	for end := 1; end < len(events); end++ {
-		if events[end].Kv.ModRevision != events[start].Kv.ModRevision {
-			r.append(batch{
-				rev:    events[start].Kv.ModRevision,
-				events: events[start:end],
+	for end := 1; end < len(items); end++ {
+		if r.revisionOf(items[end]) != r.revisionOf(items[start]) {
+			r.append(entry[T]{
+				revision: r.revisionOf(items[start]),
+				items:    items[start:end],
 			})
 			start = end
 		}
 	}
-	if start < len(events) {
-		r.append(batch{
-			rev:    events[start].Kv.ModRevision,
-			events: events[start:],
+	if start < len(items) {
+		r.append(entry[T]{
+			revision: r.revisionOf(items[start]),
+			items:    items[start:],
 		})
 	}
 }
 
-func (r *ringBuffer) append(b batch) {
-	if len(b.events) == 0 {
+func (r *ringBuffer[T]) append(entry entry[T]) {
+	if len(entry.items) == 0 {
 		return
 	}
 	if r.size == len(r.buffer) {
 		r.tail = (r.tail + 1) % len(r.buffer)
 	} else {
 		r.size++
 	}
-	r.buffer[r.head] = b
+	r.buffer[r.head] = entry
 	r.head = (r.head + 1) % len(r.buffer)
 }
 
-// AscendGreaterOrEqual iterates through batches in ascending order starting from the first batch with rev >= pivot.
-// TODO: use binary search on the ring buffer to locate the first entry >= nextRev instead of a full scan
-func (r *ringBuffer) AscendGreaterOrEqual(pivot int64, iter IterFunc) {
+// AscendGreaterOrEqual iterates through entries in ascending order starting from the first entry with revision >= pivot.
+func (r *ringBuffer[T]) AscendGreaterOrEqual(pivot int64, iter IterFunc[T]) {
 	if r.size == 0 {
 		return
 	}
 
-	for n, i := 0, r.tail; n < r.size; n, i = n+1, (i+1)%len(r.buffer) {
-		eventBatch := r.buffer[i]
-
-		if eventBatch.events == nil {
-			panic(fmt.Sprintf("ringBuffer.AscendGreaterOrEqual: unexpected nil at %d", i))
-		}
-
-		if eventBatch.rev < pivot {
-			continue
-		}
-
-		if !iter(eventBatch.rev, eventBatch.events) {
+	iterStartIndex := r.findFirstGreaterOrEqualIndex(pivot)
```

Comment:
Don't know why did you extracted this logic. Please don't mix changes in single PR.